core/src/main/scala/com/microsoft/azure/synapse/ml/explainers/ICEExplainer.scala

```diff
@@ -0,0 +1,275 @@
+
+    val dataType = df.schema(feature).dataType
+    val explodeFunc = explode(array(values.map(v => lit(v)): _*).cast(ArrayType(dataType)))
+
+    val predicted = getModel.transform(df.withColumn(feature, explodeFunc))
+    val targetCol = DatasetExtensions.findUnusedColumnName("target", predicted)
+    val dependenceCol = DatasetExtensions.findUnusedColumnName("feature__dependence", predicted)
+
+    val explainTarget = extractTarget(predicted.schema, targetClassesColumn)
+    val result = predicted.withColumn(targetCol, explainTarget)
+
+    getKind.toLowerCase match {
+      case `averageKind` =>
+        // PDP output schema: 1 row * 1 col (pdp for the given feature: feature_value -> explanations)
+        result
+          .groupBy(feature)
+          .agg(Summarizer.mean(col(targetCol)).alias(dependenceCol))
+          .agg(
+            map_from_arrays(
+              collect_list(feature),
+              collect_list(dependenceCol)
+            ).alias(outputColName)
+          )
+
+      case `individualKind` =>
+        // ICE output schema: n rows * 2 cols (idCol + ice for the given feature: map(feature_value -> explanations))
+        result
+          .groupBy(idCol)
+          .agg(
+            map_from_arrays(
+              collect_list(feature),
+              collect_list(targetCol)
+            ).alias(outputColName)
+          )
+    }
+  }
+
+  def transform(ds: Dataset[_]): DataFrame = {
+    transformSchema(ds.schema)
+    val df = ds.toDF
+    val idCol = DatasetExtensions.findUnusedColumnName("idCol", df)
+    val targetClasses = DatasetExtensions.findUnusedColumnName("targetClasses", df)
+    val dfWithId = df
+      .withColumn(idCol, monotonically_increasing_id())
+      .withColumn(targetClasses, get(targetClassesCol).map(col).getOrElse(lit(getTargetClasses)))
+
+    // Collect feature values for all features from original dataset - dfWithId
+    val (categoricalFeatures, numericFeatures) = (getCategoricalFeatures, getNumericFeatures)
+
+    // If numSamples is specified, randomly pick numSamples instances from the input dataset
+    val sampled: Dataset[Row] = get(numSamples).map(dfWithId.orderBy(rand()).limit).getOrElse(dfWithId).cache
+
+    // Collect values from the input dataframe and create dependenceDF from them
+    val calcCategoricalFunc: ICECategoricalFeature => DataFrame = {
+      f: ICECategoricalFeature =>
+        val values = collectCategoricalValues(dfWithId, f)
+        calcDependence(sampled, idCol, targetClasses, f.name, values, f.getOutputColName)
+    }
+    val calcNumericFunc: ICENumericFeature => DataFrame = {
+      f: ICENumericFeature =>
+        val values = collectSplits(dfWithId, f)
+        calcDependence(sampled, idCol, targetClasses, f.name, values, f.getOutputColName)
+    }
+
+    val dependenceDfs = (categoricalFeatures map calcCategoricalFunc) ++ (numericFeatures map calcNumericFunc)
```

Comment:
Combine both sequences then map with a case match on the Feature Class. it will be a bit cleaner and will avoid functions as values

Reply:
This still seems unresolved